### PR TITLE
DiplomatOption support in Dart

### LINKED ISSUE #246: Support optional types
We support optional opaques, but we do not support optional primitives and structs. We should; exposing it as `std::optional` in C++ and

We will likely want to do the same thing as https://github.com/rust-diplomat/diplomat/issues/245 where it's automatic when found in parameters, though we will have to require people to use DiplomatOption in struct types.

We could also make it so that you can force opaques into optionals if you really want (currently opaque+optional = nullable opaque)

Status: 
 - [x] Core support: https://github.com/rust-diplomat/diplomat/pull/558
 - [x] C/CPP: https://github.com/rust-diplomat/diplomat/pull/628
 - [ ] Dart: https://github.com/rust-diplomat/diplomat/pull/632
 - [ ] JS: Tyler will be working on this
 - [ ] Kotlin: jcrist1 may work on this

I shall consider this issue fixed when Dart, C, and CPP have landed.

### PR DESCRIPTION
**Pull request by Manishearth**
Fixes  #246

Based on https://github.com/rust-diplomat/diplomat/pull/628

### CODE REVIEW

**Comment by Manishearth** [comment]
This is reviewable, though there are no tests yet.

File: tool/templates/dart/result.dart.jinja [hunk 1]
```diff
@@ -20,4 +20,33 @@ final class {{name}} extends ffi.Struct {
 
   @ffi.Bool()
   external bool isOk;
+
+  {%- if let Some((_, ty)) = ok %}
+  factory {{name}}.ok({{ty}} val) {
```
Comment (robertbastian) [inline tool/templates/dart/result.dart.jinja:25]: nit: add a whitespace line between fields and constructors

File: tool/src/dart/mod.rs [hunk 2]
```diff
@@ -646,6 +647,10 @@ impl<'a, 'cx> TyGenContext<'a, 'cx> {
                 self.formatter.fmt_primitive_list_type(p).into()
             }
             Type::Slice(hir::Slice::Strs(..)) => "core.List<core.String>".into(),
+            Type::DiplomatOption(ref inner) => {
+                let inner = self.gen_type_name(inner);
+                format!("{inner}?").into()
```
Comment (robertbastian) [inline tool/src/dart/mod.rs:687]: use `fmt_nullable` from the formatter

File: tool/src/dart/mod.rs [hunk 3]
```diff
@@ -919,18 +935,28 @@ impl<'a, 'cx> TyGenContext<'a, 'cx> {
...
+                    };
+                    format!(
+                        "{var_name}._toDart({}Edges)",
+                        lifetime_env.fmt_lifetime(lifetime)
+                    )
+                    .into()
+                } else {
+                    format!("{var_name}._toDart([])").into()
+                }
+            }
```
Comment (robertbastian) [inline tool/src/dart/mod.rs:986]: why did you change this?

File: tool/src/dart/mod.rs [hunk 4]
```diff
@@ -438,30 +447,56 @@ impl<'a, 'cx> TyGenContext<'a, 'cx> {
...
+                        }
+                        _ => unreachable!(
+                            "Slices must produce slice ParamBorrowInfo, found {param_borrow_kind:?}"
+                        ),
                     }
-                    _ => unreachable!(
-                        "Slices must produce slice ParamBorrowInfo, found {param_borrow_kind:?}"
-                    ),
+                } else if let hir::Type::DiplomatOption(ref inner) = param_ty {
+                    alloc_name(
```
Comment (robertbastian) [inline tool/src/dart/mod.rs:482]: not a fan of the recursion. Can you instead match on `DiplomatOption(Struct)` in addition to `Struct`?

File: tool/src/dart/mod.rs (hunk 3 above)
Comment (Manishearth) [reply tool/src/dart/mod.rs:986]: type inference was failing on the lack of `into()`, apparently it was only working when the last branch is the one with the String.

File: tool/src/dart/mod.rs (hunk 4 above)
Comment (Manishearth) [reply tool/src/dart/mod.rs:482]: This would subtly break if we ever added option-slice support, though?

File: tool/src/dart/mod.rs (hunk 3 above)
Comment (robertbastian) [reply tool/src/dart/mod.rs:986]: That's weird, both branches are strongly typed, it shouldn't make a difference whether the `into` is on the branches or outside

File: tool/src/dart/mod.rs (hunk 4 above)
Comment (robertbastian) [reply tool/src/dart/mod.rs:482]: You can match on `DiplomatOption`, and the let-bind the struct with an unreachable else case.

**Review by robertbastian — COMMENTED** [review]
.

File: tool/src/dart/mod.rs (hunk 4 above)
Comment (robertbastian) [reply tool/src/dart/mod.rs:482]: Or have a `DiplomatOption(Struct)` case with the `Struct` case, and a `DiplomatOption(..)` case that is unreachable.

Comment (Manishearth) [reply tool/src/dart/mod.rs:482]: note that `box` patterns are unstable so you cannot use the same match arm. I cannot do `Struct(s) | DiplomatOption(Struct(s))`, so either way I need _some_ recursion or at least a more scope function for structs.

File: tool/src/dart/mod.rs (hunk 3 above)
Comment (Manishearth) [reply tool/src/dart/mod.rs:986]: Rust type inference gets really weird around generics like Into. It has some smart heuristics but they often fail like this.

`.into()` is not "strongly typed" there's a lot of inference there, and IIRC there's Into-specific magic that makes Into work okayishly but it's brittle

Comment (robertbastian) [reply tool/src/dart/mod.rs:986]: I mean `String` is strongly typed. Before, the if-else had type `String`, which then got `into`d. Now each branch `into`s the `String` separately and that's somehow easier for the compiler!?

File: tool/src/dart/mod.rs (hunk 4 above)
Comment (robertbastian) [reply tool/src/dart/mod.rs:482]: ugh

Comment (Manishearth) [reply tool/src/dart/mod.rs:482]: yeah, I started writing it this way first and quickly remembered box patterns were unstable.

**Review by robertbastian — COMMENTED** [review]
Can you fix the white space diffs in the result fields?

**Comment by Manishearth** [comment]
I thought I fixed them?

File: feature_tests/dart/lib/src/lib.g.dart [hunk 5]
```diff
@@ -125,12 +128,25 @@ final class _ResultDoubleVoid extends ffi.Struct {
...
+    return struct;
+  }
 }
 
 final class _ResultInt32OpaqueUnion extends ffi.Union {
   @ffi.Int32()
-  external int ok;
 
+  external int ok;
   external ffi.Pointer<ffi.Opaque> err;
```
Comment (robertbastian) [inline feature_tests/dart/lib/src/lib.g.dart:150]: there should not be a blank line between the annotation and the field, and there should be one between the fields

Comment (Manishearth) [reply feature_tests/dart/lib/src/lib.g.dart]: Ah, found it

File: feature_tests/dart/lib/src/lib.g.dart [hunk 6]
```diff
@@ -176,11 +229,24 @@ final class _ResultOpaqueErrorStructFfi extends ffi.Struct {
...
+    final struct = ffi.Struct.create<_ResultOpaqueErrorStructFfi>();
+    struct.isOk = false;
+    struct.union.err = val;
+    return struct;
+  }
 }
 
 final class _ResultOpaqueInt32Union extends ffi.Union {
   external ffi.Pointer<ffi.Opaque> ok;
-
```
Comment (robertbastian) [inline feature_tests/dart/lib/src/lib.g.dart:183]: ...

Comment (Manishearth) [reply feature_tests/dart/lib/src/lib.g.dart:183]: this should be fine? I'd rather have that whitespace be adjacent anyway, instead of complicating the template with a manual `if is_ok() & is_err()` subst _just for whitespace_.

Comment (robertbastian) [reply feature_tests/dart/lib/src/lib.g.dart:183]: The template should add an empty line after each field, there shouldn't need to be any special casing.

Comment (Manishearth) [reply feature_tests/dart/lib/src/lib.g.dart:183]: I can't see a way to do this without either:

 - dedenting some of the `{%`s (because removing the `-` also reintroduces the indentation)
 - adding some kind of `if` somewhere

Comment (Manishearth) [reply feature_tests/dart/lib/src/lib.g.dart]: I pushed the dedent solution. It works, I'm not happy with it

**Review by robertbastian — APPROVED** [review]
